backend/src/test/java/fk/prof/backend/LeaderElectionTest.java

```diff
@@ -0,0 +1,244 @@
+    curatorClient.start();
+    curatorClient.blockUntilConnected(10, TimeUnit.SECONDS);
+  }
+
+  @After
+  public void tearDown(TestContext context) throws IOException {
+    System.out.println("Tearing down");
+    VertxManager.close(vertx).setHandler(result -> {
+      System.out.println("Vertx shutdown");
+      curatorClient.close();
+      try {
+        testingServer.close();
+      } catch (IOException ex) {
+      }
+      if (result.failed()) {
+        context.fail(result.cause());
+      }
+    });
+  }
+
+  @Test(timeout = 20000)
+  public void leaderTaskTriggerOnLeaderElection(TestContext testContext) throws InterruptedException {
+    vertx = vertxConfig != null ? Vertx.vertx(new VertxOptions(vertxConfig)) : Vertx.vertx();
+    CountDownLatch latch = new CountDownLatch(1);
+    Runnable leaderElectedTask = () -> {
+      latch.countDown();
+    };
+    LeaderDiscoveryStore leaderDiscoveryStore = VertxManager.getDefaultLeaderDiscoveryStore(vertx);
+
+    Thread.sleep(1000);
+    VertxManager.deployLeaderElectionWorkerVerticles(
+        vertx,
+        leaderElectionDeploymentOptions,
+        curatorClient,
+        leaderElectedTask,
+        leaderDiscoveryStore
+    );
+
+    boolean released = latch.await(10, TimeUnit.SECONDS);
+    if (!released) {
+      testContext.fail("Latch timed out but leader election task was not run");
+    }
+  }
+
+  @Test(timeout = 20000)
+  public void leaderDiscoveryUpdateOnLeaderElection(TestContext testContext) throws InterruptedException {
+    vertx = vertxConfig != null ? Vertx.vertx(new VertxOptions(vertxConfig)) : Vertx.vertx();
+    CountDownLatch latch = new CountDownLatch(1);
+    Runnable leaderElectedTask = VertxManager.getDefaultLeaderElectedTask(vertx, true, null, false, null, null, null);
+    LeaderDiscoveryStore leaderDiscoveryStore = new LeaderDiscoveryStore() {
+      private String address = null;
+      private boolean self = false;
+
+      @Override
+      public void setLeaderIPAddress(String ipAddress) {
+        address = ipAddress;
+        self = ipAddress != null && ipAddress.equals(IPAddressUtil.getIPAddressAsString());
+        if (address != null) {
+          latch.countDown();
+        }
+      }
+
+      @Override
+      public String getLeaderIPAddress() {
+        return address;
+      }
+
+      @Override
+      public boolean isLeader() {
+        return self;
+      }
+    };
+
+    Thread.sleep(1000);
+    VertxManager.deployLeaderElectionWorkerVerticles(
+        vertx,
+        leaderElectionDeploymentOptions,
+        curatorClient,
+        leaderElectedTask,
+        leaderDiscoveryStore
+    );
+
+    boolean released = latch.await(10, TimeUnit.SECONDS);
+    if (!released) {
+      testContext.fail("Latch timed out but leader discovery store was not updated with leader address");
+    } else {
+      testContext.assertEquals(IPAddressUtil.getIPAddressAsString(), leaderDiscoveryStore.getLeaderIPAddress());
```

Comment:
vertx has this concept of async.complete. Unless that is called, test is not successful and times out. This is to make testing async code easier. You only need to place async.complete call in your happy path (at the deepest nested assert, usually).